src/kernel/binary_reduce.cc

```diff
@@ -0,0 +1,506 @@
+      oss << data[i] << " ";
+    }
+  }
+  oss << "]";
+  return oss.str();
+}
+
+inline void CheckCtx(
+    const DLContext& ctx,
+    const std::vector<NDArray>& arrays,
+    const std::vector<std::string>& names) {
+  for (size_t i = 0; i < arrays.size(); ++i) {
+    if (utils::IsNoneArray(arrays[i]))
+      continue;
+    CHECK_EQ(ctx, arrays[i]->ctx)
+      << "Expected device context " << ctx << ". But got "
+      << arrays[i]->ctx << " for " << names[i] << ".";
+  }
+}
+
+inline void CheckIdArray(
+    const uint8_t bits,
+    const std::vector<NDArray>& arrays,
+    const std::vector<std::string>& names) {
+  for (size_t i = 0; i < arrays.size(); ++i) {
+    if (utils::IsNoneArray(arrays[i]))
+      continue;
+    CHECK(arrays[i]->dtype.code == kDLInt);
+    CHECK_EQ(arrays[i]->ndim, 1);
+    CHECK_EQ(bits, arrays[i]->dtype.bits)
+      << "Expected " << bits << " integer array. But got "
+      << arrays[i]->dtype.bits << " for " << names[i] << ".";
+  }
+}
+
+// Return true if the operator is commutative and lhs and rhs need
+// to be switched. For example, Add(kDst, kSrc) needs to be changed
+// to Add(kSrc, kDst)
+inline bool NeedSwitchOrder(const std::string& op,
+    binary_op::Target lhs, binary_op::Target rhs) {
+  CHECK_NE(lhs, rhs);
+  return (op == binary_op::kAdd || op == binary_op::kMul)
+    && lhs > rhs;
+}
+
+}  // namespace
+
+
+std::vector<int64_t> InferBinaryFeatureShape(
+    NDArray lhs,
+    NDArray rhs) {
+  return CalcBcastInfo(lhs, rhs).real_out_shape;
+}
+
+DGL_REGISTER_GLOBAL("kernel._CAPI_DGLKernelInferBinaryFeatureShape")
+.set_body([] (DGLArgs args, DGLRetValue* rv) {
+    NDArray lhs = args[0];
+    NDArray rhs = args[1];
+    const auto& shape = InferBinaryFeatureShape(lhs, rhs);
+    const int64_t len = shape.size();
+    NDArray ret = NDArray::Empty(
+        {len}, DLDataType{kDLInt, 64, 1}, DLContext{kDLCPU, 0});
+    int64_t* ret_data = static_cast<int64_t*>(ret->data);
+    std::copy(shape.begin(), shape.end(), ret_data);
+    *rv = ret;
+  });
+
+void BinaryOpReduce(
+    const std::string& reducer,
+    const std::string& op,
+    const ImmutableGraph* graph,
+    binary_op::Target lhs, binary_op::Target rhs,
+    NDArray lhs_data, NDArray rhs_data,
+    NDArray out_data,
+    NDArray lhs_mapping, NDArray rhs_mapping,
+    NDArray out_mapping) {
+  const auto& ctx = graph->Context();
+  // sanity check
+  CheckCtx(ctx,
+      {lhs_data, rhs_data, out_data, lhs_mapping, rhs_mapping, out_mapping},
+      {"lhs_data", "rhs_data", "out_data", "lhs_mapping", "rhs_mapping", "out_mapping"});
+  CheckIdArray(graph->NumBits(),
+      {lhs_mapping, rhs_mapping, out_mapping},
+      {"lhs_mapping", "rhs_mapping", "out_mapping"});
+  // Switch order for commutative operation
+  if (NeedSwitchOrder(op, lhs, rhs)) {
+    BinaryOpReduce(reducer, op, graph,
+        rhs, lhs, rhs_data, lhs_data, out_data,
+        rhs_mapping, lhs_mapping, out_mapping);
+  } else {
+    if (HasBcast(lhs_data, rhs_data)) {
```

Comment:
is it necessary to support broadcasting? what is the use case?

Reply:
Multi-head graph attention network. The node feature shape is (H, D) and the edge feature shape is (H, 1).